Content.Server/Bed/Sleep/SleepingSystem.cs

```diff
@@ -0,0 +1,224 @@
+
+            args.Verbs.Add(verb);
+        }
+
+        /// <summary>
+        /// When you click on a sleeping person with an empty hand, try to wake them.
+        /// </summary>
+        private void OnInteractHand(EntityUid uid, SleepingComponent component, InteractHandEvent args)
+        {
+            args.Handled = true;
+            TryWaking(args.Target, user: args.User);
+        }
+
+        private void OnExamined(EntityUid uid, SleepingComponent component, ExaminedEvent args)
+        {
+            if (args.IsInDetailsRange)
+            {
+                args.PushMarkup(Loc.GetString("sleep-examined", ("target", Identity.Entity(uid, EntityManager))));
+            }
+        }
+
+        /// <summary>
+        /// Try sleeping. Only mobs can sleep.
+        /// </summary>
+        public bool TrySleeping(EntityUid uid)
+        {
+            if (!HasComp<MobStateComponent>(uid))
+                return false;
+
+            if (_prototypeManager.TryIndex<InstantActionPrototype>("Sleep", out var sleepAction))
+                _actionsSystem.RemoveAction(uid, sleepAction);
+
+            EnsureComp<SleepingComponent>(uid);
+            return true;
+        }
+
+        /// <summary>
+        /// This adds forced sleep that will not wake up on damage.
+        /// Lots of chems and disease effects use these.
+        /// You can always set it to 0 or a negative value too, I guess.
+        /// </summary>
+        public bool AddForcedSleepingTime(EntityUid uid, float secondsToAdd)
+        {
+            if (!HasComp<MobStateComponent>(uid))
+                return false;
+
+            EnsureComp<SleepingComponent>(uid);
+            var forced = EnsureComp<ForcedSleepingComponent>(uid);
+            forced.Accumulator -= secondsToAdd;
+            return true;
+        }
+
+        /// <summary>
+        /// Try to wake up.
+        /// </summary>
+        public bool TryWaking(EntityUid uid, bool force = false, EntityUid? user = null)
+        {
+
+            if (!force && HasComp<ForcedSleepingComponent>(uid))
+            {
+                if (user != null)
+                {
+                    SoundSystem.Play("/Audio/Effects/thudswoosh.ogg", Filter.Pvs(uid), uid, AudioHelpers.WithVariation(0.05f, _robustRandom));
+                    _popupSystem.PopupEntity(Loc.GetString("wake-other-failure", ("target", Identity.Entity(uid, EntityManager))), uid, Filter.Entities(user.Value), Shared.Popups.PopupType.SmallCaution);
+                }
+                return false;
+            }
+
+            if (user != null)
+            {
+                SoundSystem.Play("/Audio/Effects/thudswoosh.ogg", Filter.Pvs(uid), uid, AudioHelpers.WithVariation(0.05f, _robustRandom));
+                _popupSystem.PopupEntity(Loc.GetString("wake-other-success", ("target", Identity.Entity(uid, EntityManager))), uid, Filter.Entities(user.Value));
+            }
+            RemComp<SleepingComponent>(uid);
+            return true;
```

Comment:
this is kind of duplicated with the knockdown code for the same thing isnt it?

Reply:
duplicated because it plays a sound? The other stuff is different and this actually works unlike the knockdown one